File: 17-03-2025_Subset_sum_problem.cpp
bool isSubsetSum(vector<int>& arr, int sum) {
    int n = arr.size();
    
    vector<vector<int>> dp (n+1, vector<int>(sum+1, 0));
    
    for(int i = 0; i <= n; i++){
        dp[i][0] = 1;
    }
    
    for(int i = 1; i <= n; i++){
        for(int j = 1; j <= sum; j++){
            if(arr[i-1] > j)
                dp[i][j] = dp[i-1][j];
            else{
                dp[i][j] = (dp[i-1][j] | dp[i-1][j - arr[i-1]]);
            }
        }
    }
    
    return (bool)dp[n][sum];
}

File: 17-03-2025_Subset_sum_problem.java
static Boolean isSubsetSum(int arr[], int sum) {
    // code here
    int n = arr.length;
    boolean[][] dp = new boolean[n + 1][sum + 1];

    for (int i = 0; i <= n; i++) {
        dp[i][0] = true; // Base case: sum = 0 is always possible with an empty subset
    }

    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= sum; j++) {
            if (arr[i - 1] > j)
                dp[i][j] = dp[i - 1][j]; // Exclude the current element
            else
                dp[i][j] = dp[i - 1][j] || dp[i - 1][j - arr[i - 1]]; // Include or exclude the element
        }
    }

    return dp[n][sum];
}
